### ISSUE TITLE
[ttnn/data_movement] transpose/permute WH on RM + HEIGHT-sharded in → WIDTH-sharded out: silent all-zero output (num_hw_blocks_per_core=0)

### ISSUE THREAD

**Issue by mouliraj-mcw** [tenstorrent/tt-metal#55582]
## Summary

`ttnn.transpose(x, -2, -1)` (and `ttnn.permute(x, (0,1,3,2))` which routes through it) silently returns wrong data when the input is **ROW_MAJOR + HEIGHT_SHARDED** and the output is (either inherited or explicitly requested) **WIDTH_SHARDED**. No error, no fatal — just an output tensor with PCC ≈ 0 against torch.

Discovered via the TM Layout/Sharding audit under #25529 (§4.1). Not covered by #51214 (static-audit list).

## Repro (wormhole_b0, `main @ 6503ed3d720`)

```python
import torch, ttnn
def hs(cores, sh, sw):
    grid = ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0,0), ttnn.CoreCoord(cores-1,0))})
    return ttnn.MemoryConfig(
        ttnn.TensorMemoryLayout.HEIGHT_SHARDED, ttnn.BufferType.L1,
        ttnn.ShardSpec(grid, (sh, sw), ttnn.ShardOrientation.ROW_MAJOR),
    )
d = ttnn.open_device(device_id=0)
torch.manual_seed(0)
t = torch.randn(1, 1, 64, 128, dtype=torch.bfloat16)
for cores in (2, 4, 8):
    x = ttnn.from_torch(t, layout=ttnn.ROW_MAJOR_LAYOUT, device=d,
                        memory_config=hs(cores, 64 // cores, 128))
    o_bad = ttnn.transpose(x, -2, -1)                                # inherits WIDTH_SHARDED
    o_ok  = ttnn.transpose(x, -2, -1, memory_config=ttnn.DRAM_MEMORY_CONFIG)
    print(cores, "bad", torch.corrcoef(torch.stack([
        ttnn.to_torch(o_bad).float().flatten(),
        t.transpose(-2,-1).contiguous().float().flatten()]))[0,1].item())
```

Result:

| cores | inherited-out PCC | DRAM_IL PCC |
|---|---|---|
| 2 | **0.0118** 🐞 | 1.0000 ✅ |
| 4 | **0.0097** 🐞 | 1.0000 ✅ |
| 8 | **-0.0003** 🐞 | 1.0000 ✅ |

`ttnn.permute(x, (0,1,3,2))` reproduces (PCC 0.0037 / 0.0017 / nan) since `permute.cpp:99-101` delegates to `ttnn::transpose(-2, -1)`.

## Root cause

Routing in `transpose_device_operation.cpp:130-131`:

```
bool use_sharded_wh =
    native && ((input_width_and_height_fully_in_shard && output_width_and_height_fully_in_shard) ||
               (N == 1 && C == 1 && input_height_sharded && output_width_sharded));
```

The second disjunct admits cases where `shard_shape[0] < H` (i.e. the H dim is split across cores) and, on RM, forwards to `TransposeWHShardedRMProgramFactory`. That factory computes:

```
transpose_wh_sharded_rm_program_factory.cpp:76
uint32_t num_hw_blocks_per_core = shard_height / H;   // truncates to 0 when shard_height < H
```

`num_hw_blocks_per_core = 0` is passed as compile-time arg 0 to reader / writer / compute kernels (lines 170, 191, 213); every per-core outer loop is `for (n = 0; n < num_hw_blocks_per_core; …)` and never executes, leaving the output CB unpopulated. Matches the sibling `transpose_hc_sharded_program_factory.cpp:124` which already clamps: `shard_height / H > 0 ? shard_height / H : 1`.

## Suggested fix

Either

1. Tighten routing — gate the RM branch of `use_sharded_wh` on `input_width_and_height_fully_in_shard && output_width_and_height_fully_in_shard`, so the H-split case falls through to `TransposeWHProgramFactory` (verified correct at PCC 1.0 for the same input in this repro).
2. Fix the factory to handle `shard_height < H` (bigger surgery; unnecessary if #1 lands since `TransposeWHProgramFactory` already covers this shape).

Preferred: #1.

## Test gap

`tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_tranpose.py`'s RM matrix (lines 423-500) has entries for `block→interleaved`, `width→interleaved`, `interleaved→block`, `interleaved→width`, `block→block`, `height→interleaved` — no `height→width` RM cell. That is the exact cell that returns garbage.

## Related

- Parent tracker: #25529 (audit), #31716 (universal-input TMs)
- Not in #51214 defect list
- Same audit calls out three sibling silent-corruption bugs: clone DRAM_H-sharded (§4.2), move sharded-out (§4.3), untilize_with_unpadding BLOCK/ND-sharded-out (§4.4) — separate issues.
